src/storage/txn/flow_controller.rs

```diff
@@ -0,0 +1,934 @@
+//   * better control on the throttle rate to avoid QPS drop under heavy write
+//
+// When write stall happens, the max speed of write rate max_delayed_write_rate
+// is limited to 16MB/s by default which doesn't take real disk ability into
+// account. It may underestimate the disk's throughout that 16MB/s is too small
+// at once, causing a very large jitter on the write duration.
+// Also, it decreases the delayed write rate further if the factors still exceed
+// the threshold. So under heavy write load, the write rate may be throttled to
+// a very low rate from time to time, causing QPS drop eventually.
+//
+// This main idea of the flow controller is to throttle at a steady write rate
+// so that the number of L0 keeps around the threshold. When it falls below the
+// threshold, the throttle state wouldn't exit right away. Instead, it may keep
+// or increase the throttle speed depending on some statistics.
+// How can we decide the throttle speed? As we can imagine, the consumption
+// ability of L0 wouldn't change dramatically corresponding to the ability of
+// hardware. So we can record the flush flow when reach the threshold as target
+// flow, and increase or decrease the throttle speed based on whether current
+// flush flow is smaller/larger than target flow.
+
+// For compaction pending bytes, we use discardable ratio...(TBD)
+pub struct FlowController {
+    discard_ratio: Arc<AtomicU64>,
+    limiter: Arc<Limiter>,
+    tx: Sender<bool>,
+    handle: Option<std::thread::JoinHandle<()>>,
+}
+
+impl Drop for FlowController {
+    fn drop(&mut self) {
+        let h = self.handle.take();
+        if h.is_none() {
+            return;
+        }
+
+        if let Err(e) = self.tx.send(true) {
+            error!("send quit message for time monitor worker failed"; "err" => ?e);
+            return;
+        }
+
+        if let Err(e) = h.unwrap().join() {
+            error!("join time monitor worker failed"; "err" => ?e);
+            return;
+        }
+    }
+}
+
+impl FlowController {
+    pub fn new<E: Engine>(
+        config: &Config,
+        engine: E,
+        flow_info_receiver: Option<Receiver<FlowInfo>>,
+    ) -> Self {
+        let limiter = Arc::new(Limiter::new(INFINITY));
+        let discard_ratio = Arc::new(AtomicU64::new(0));
+        let checker = FlowChecker::new(config, engine, discard_ratio.clone(), limiter.clone());
+        let (tx, rx) = mpsc::channel();
+        let handle = if config.disable_write_stall {
+            Some(checker.start(rx, flow_info_receiver.unwrap()))
+        } else {
+            None
+        };
+        Self {
+            discard_ratio,
+            limiter,
+            tx,
+            handle,
+        }
+    }
+
+    pub fn should_drop(&self) -> bool {
+        let ratio = self.discard_ratio.load(Ordering::Relaxed) as f64 / RATIO_PRECISION;
+        let mut rng = rand::thread_rng();
+        rng.gen::<f64>() < ratio
```

Comment:
Could we generate rand int in [0, RATIO_PRECISION] instead? BTW maybe rename it to `RATIO_SCALE_FACTOR`.